pkg/kstatus/observe/reader/cached_reader.go

```diff
@@ -0,0 +1,207 @@
+package reader
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"sigs.k8s.io/cli-utils/pkg/kstatus/wait"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// genGroupKinds captures the generated resources for a given GroupKind. This is used
+// to expand the list GVK and namespace combinations that needs to be synced in the cache.
+// TODO: These should probably be defined in the observers rather than here.
+var genGroupKinds = map[schema.GroupKind][]schema.GroupKind{
+	schema.GroupKind{Group: "apps", Kind: "Deployment"}: { //nolint:gofmt
+		{
+			Group: "apps",
+			Kind:  "ReplicaSet",
+		},
+	},
+	schema.GroupKind{Group: "apps", Kind: "ReplicaSet"}: { //nolint:gofmt
+		{
+			Group: "",
+			Kind:  "Pod",
+		},
+	},
+	schema.GroupKind{Group: "apps", Kind: "StatefulSet"}: { //nolint:gofmt
+		{
+			Group: "",
+			Kind:  "Pod",
+		},
+	},
+}
+
+// NewCachedObserverReader returns a new instance of the ObserverReader.
+func NewCachedObserverReader(reader client.Reader, mapper meta.RESTMapper, identifiers []wait.ResourceIdentifier) (*CachedObserverReader, error) {
+	gvkNamespaceSet := newGnSet()
+	for _, id := range identifiers {
+		// For every identifier, add the GroupVersionKind and namespace combination to the gvkNamespaceSet and
+		// check the genGroupKinds map for any generated resources that also should be included.
+		err := buildGvkNamespaceSet(mapper, []schema.GroupKind{id.GroupKind}, id.Namespace, gvkNamespaceSet)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return &CachedObserverReader{
+		reader: reader,
+		mapper: mapper,
+		gns:    gvkNamespaceSet.gvkNamespaces,
+	}, nil
+}
+
+func buildGvkNamespaceSet(mapper meta.RESTMapper, gks []schema.GroupKind, namespace string, gvkNamespaceSet *gvkNamespaceSet) error {
+	for _, gk := range gks {
+		mapping, err := mapper.RESTMapping(gk)
+		if err != nil {
+			return err
+		}
+		gvkNamespaceSet.add(gvkNamespace{
+			GVK:       mapping.GroupVersionKind,
+			Namespace: namespace,
+		})
+		genGKs, found := genGroupKinds[gk]
+		if found {
+			err := buildGvkNamespaceSet(mapper, genGKs, namespace, gvkNamespaceSet)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+type gvkNamespaceSet struct {
+	gvkNamespaces []gvkNamespace
+	seen          map[gvkNamespace]bool
+}
+
+func newGnSet() *gvkNamespaceSet {
+	return &gvkNamespaceSet{
+		gvkNamespaces: make([]gvkNamespace, 0),
+		seen:          make(map[gvkNamespace]bool),
+	}
+}
+
+func (g *gvkNamespaceSet) add(gn gvkNamespace) {
+	if _, found := g.seen[gn]; !found {
+		g.gvkNamespaces = append(g.gvkNamespaces, gn)
+		g.seen[gn] = true
+	}
+}
+
+// CachedObserverReader is an implementation of the ObserverReader interface that will
```

Comment:
An `ObserverReader` sounds like something that (oddly)  _reads an observer_.

But the interface is more clearly a `ClusterReader` or `SyncableClusterReader` or ...

It's clearly a `client.Reader` adapter.  Naming it after the value added by the adaptation would be reasonable. But the value add at the _interface_ level isn't clear.

I know the interface is already in, but i'm objecting anyway :)

Here we have `CachedObserverReader`, which sounds like a reader that is kept in a cache.

Suggest the name:  `CachingClusterReader`.

Reply:
Agree. I have renamed the caching implementation to `CachingClusterReader` and the implementation which just wraps `client.Reader` to `DirectClusterReader`.
I will do a follow up PR to rename the interface.